design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,296 @@
+* The client will initially support execution and streaming of queries (both push and pull), inserting of rows into streams, DDL operations and admin operations
+such as list and describe.
+* The client will support a reactive / async interaction style using Java CompletableFuture and Reactive Streams / JDK9 Flow API (or a copy thereof if we cannot upgrade from Java 8)
+* The client will also support a direct / synchronous interaction style.
+* The server API will be very simple, based on HTTP2 with a simple text/json based encoding therefore it will be very simple to use directly from vanilla
+HTTP2 clients if a ksqldb client is not available
+
+### The Server API
+
+We will create a new, simple HTTP2 API for streaming query results from server to client and
+for streaming inserts from client to server. We will use a simple text/JSON encoding for data.
+Please note, this is not a REST API, it's a streaming HTTP API. The API does not follow REST principles. REST is inherently designed for request/response (RPC)
+style interactions, not streaming.
+
+The API will have the following characteristics:
+
+* Multiplexed (because of HTTP2 multiplexing)
+* Back-pressure (because of HTTP2 flow control)
+* Text based so easy to use and view results using command line tools such as curl
+* Can be used from any vanilla HTTP2 client for any language
+* Simple newline based delimitation so easy to parse results on client side
+* JSON encoding of data so very easy to parse as most languages have good JSON support
+* CORS support
+* HTTP basic auth support
+* TLS
+
+#### Query streaming
+
+The request method will be a POST.
+
+Requests will be made to a specific URL, e.g. "/query-stream" (this can be configurable)
+
+The body of the request is a JSON object UTF-8 encoded as text, containing the arguments for the
+operation (newlines have been added here for the sake of clarity but the real JSON must not contain newlines)
+
+````
+{
+"query": "select * from foo", <----- the SQL of the query to execute
+"push": true,                 <----- if true then push query, else pull query
+"limit": 10                   <----  If specified return at most this many rows,
+"properties": {               <----- Optional properties for the query
+    "prop1": "val1",
+    "prop2": "val2"
+   }
+}
+
+````
+
+Please note the parameters are not necessarily exhaustive. The description here is an outline not a detailed
+low level design. The low level design will evolve during development.
+
+In the case of a successful query 
+
+````
+{
+"query-id", "xyz123",                          <---- unique ID of the query, used when terminating the query
+"columns":["col", "col2", "col3"],             <---- the names of the columns
+"column_types":["BIGINT", "STRING", "BOOLEAN"] <---- The types of the columns
+"row_count": 101                               <---- The number of rows - only set in case of pull query
+}
+````
+
+Followed by zero or more JSON arrays:
+
+````
+[123, "blah", true]
```

Comment:
I understand that we write the arrays outside of the json (IIUC) above, because when we stream results back, we don't know how many rows there are to return..

But in cases we do, e.g pull queries, could we support an encoding where this array is within the json above and the overall response is a valid json (as it is today)? The advantage of doing that would be 

- Opens up integration with lots of systems/tools that can do something useful by calling an endpoint and getting a json. cc @AlanConfluent @vpapavas  who have built a ksqlDB/Grafana integration, that uses pull queries results returns as json.. I think the new encoding would break such an use-case? 
- Even for push queries, IMO returning a full json is great because, then a client can fully consume a small table with a `LIMIT` clause added.. Again we had a real use-case for this in the grafana integration, where we wanted to write a push query to consume all the new metric names (not actual metric events) ...

Reply:
I agree that it's probably preferable if we have a single response that's parsable as valid json.  Otherwise, you have to do some parsing yourself (even if separated by just \n), which isn't preferable.  Also, some frameworks, including grafana make assumptions about a single parsable json response.  There might be workarounds, but this would probably go against a lot of clients out there.  Best to avoid this.  

Also, I think there's certainly something nice about getting just one format that has all of the column information and results in a single response.  Consistency between pull and push would also be nice, so that handling logic could be shared, even if it means that results are a bit more verbose.

Reply:
> I understand that we write the arrays outside of the json (IIUC) above, because when we stream results back, we don't know how many rows there are to return..

This is more about we don't want to force users to have to parse huge JSON objects into memory in order to do something useful with them. Yes, streaming JSON parsers do exist but we shouldn't assume they do on all target client languages or mandate users to use them. Async (non blocking) streaming JSON parsers are even more rare.

> 
> But in cases we do, e.g pull queries, could we support an encoding where this array is within the json above and the overall response is a valid json (as it is today)? The advantage of doing that would be

We shouldn't assume that pull query results will be small, they could be huge. However, if you can provide a limit clause and thus provide a hard limit on the results, then returning the results as a single JSON object can make sense.

We can easily enable this by providing the content-type in the accepts header of the request. I.e. if you want the results a single JSON object provide content-type="text/json", we can define a new content-type for our newline delimited sequence of JSON elements (which should be the default). The server can then fail requests with content-type="text/json" if a limit clause hasn't been provided.

> 
> * Opens up integration with lots of systems/tools that can do something useful by calling an endpoint and getting a json. cc @AlanConfluent @vpapavas  who have built a ksqlDB/Grafana integration, that uses pull queries results returns as json.. I think the new encoding would break such an use-case?

All of the proposals for the new API (custom protocol, gRPC) would break this. Custom protocol + gRPC are definitely not JSON ;)

So I'm working on the assumption that you would have to fix Grafana anyway. Presumably the tool allows you to provide some kind of plugin that you can code to work with some kind of custom API? If not, it seems rather limited.

I don't think it's generally a good approach to constrain our API based on the requirements of one particular tool which we happen to use internally.

Having said that, if Grafana does require a single JSON, then the new API can support this using an accepts header as mentioned above :)

> * Even for push queries, IMO returning a full json is great because, then a client can fully consume a small table with a `LIMIT` clause added.. Again we had a real use-case for this in the grafana integration, where we wanted to write a push query to consume all the new metric names (not actual metric events) ...

As long as limit is specified I see no issue with that.

Reply:
> I agree that it's probably preferable if we have a single response that's parsable as valid json. Otherwise, you have to do some parsing yourself (even if separated by just \n), which isn't preferable. Also, some frameworks, including grafana make assumptions about a single parsable json response. There might be workarounds, but this would probably go against a lot of clients out there. Best to avoid this.

Please see reply to Vinoth who made a similar observation.

Reply:
Another advantage of not providing results as a single json element. Quite often in the world of streaming you're just piping streams from one place to another. A concrete use case for this is piping the output of a query to a websocket or SSE so it can be sent to a browser (this will probably be a common use case for ksql based apps).
Ih this case you don't need to look at the individual JSON results you just want to pick up each individual row as bytes and write them as bytes. Using a streaming JSON parser you'd end up deserialising the JSON from bytes only to deserialise it again back to bytes before writing it to the websocket. This is very inefficient. Using some kind of framing outside the JSON (in this case delimited, but it also could be length prefixed) allows you to pick out the elements much more easily and efficiently.

Reply:
> I agree that it's probably preferable if we have a single response that's parsable as valid json. Otherwise, you have to do some parsing yourself (even if separated by just \n), 

It's generally going to be a lot simpler to parse on newlines in <insert language here> then to assume the presence of a streaming JSON parser in <insert language here> and figure out how to use it to extract the individual elements.

I do accept there are some use cases where you know the results are not going to be large (and you use limit) and you want to parse them all as a single JSON object, but this is a streaming API so we can't make that assumption in general.

Reply:
> if you want the results a single JSON object provide content-type="text/json", we can define a new content-type for our newline delimited sequence of JSON elements (which should be the default)

nit: the single-object content type should be `application/json` to match [RFC 4627](https://www.ietf.org/rfc/rfc4627.txt).

> we can define a new content-type for our newline delimited sequence of JSON elements (which should be the default)

For reference, the custom content type supported by the [existing REST API](https://github.com/confluentinc/ksql/blob/230868683a4c20c656d9f00226a300963944c0bc/docs/developer-guide/api.rst#content-types) is `application/vnd.ksql.v1+json`.

Reply:
> A concrete use case for this is piping the output of a query to a websocket or SSE

Speaking of this, what are the downsides of using SSE as the transport directly, for streaming query responses? A response of Content-Type `text/event-stream` corresponding to the example under the KLIP's `Query streaming` heading could look like:

```
event: query-success
data: { "query-id", "xyz123", "columns":["col", "col2", "col3"],  "column_types":["BIGINT", "STRING", "BOOLEAN"], "row_count": 101 }

event: query-data
data: [123, "blah", true]

data: [432, "foo", true]

data: [765, "whatever", false]
```

Reply:
I think SSE is a valid alternate API we can provide that some users might prefer. One thing about the new API implementation is the large majority of the work is in setting up the infrastructure, getting the new Vert.x based server plugged-in and integrating it with the current backend. Once that is done exposing the query streams in different formats is relatively straightforward - it's pretty much just writing the output to the response in a different way. I'd recommend we initially ship with the formats as described in the KLIP and we can consider adding SSE support at a later date (it wouldn't be hard).